Some Spot are created without Tag

Some spot requests are fulfilled, but no 'launched-for-asg' is been assigned to this instance.

03:08:27
autoscaling.go:443: production-ecs-cluster Created spot instance request sir-dbqr5wjj
03:08:27
autoscaling.go:483: production-ecs-cluster Failed to create tags for the spot instance request InvalidSpotInstanceRequestID.NotFound: The spot instance request ID 'sir-dbqr5wjj' does not exist
03:08:28
autoscaling.go:335: production-ecs-cluster Refreshed details for sir-dbqr5wjj {
03:08:28
SpotInstanceRequestId: "sir-dbqr5wjj",
03:08:33
autoscaling.go:335: production-ecs-cluster Refreshed details for sir-dbqr5wjj {
03:08:33
SpotInstanceRequestId: "sir-dbqr5wjj",
03:08:38
autoscaling.go:335: production-ecs-cluster Refreshed details for sir-dbqr5wjj {
03:08:38
SpotInstanceRequestId: "sir-dbqr5wjj",
03:08:43
autoscaling.go:335: production-ecs-cluster Refreshed details for sir-dbqr5wjj {
03:08:43
SpotInstanceRequestId: "sir-dbqr5wjj",
03:08:48
autoscaling.go:335: production-ecs-cluster Refreshed details for sir-dbqr5wjj {
03:08:48
SpotInstanceRequestId: "sir-dbqr5wjj",
03:08:53
autoscaling.go:335: production-ecs-cluster Refreshed details for sir-dbqr5wjj {
03:08:53
SpotInstanceRequestId: "sir-dbqr5wjj",
03:08:58
autoscaling.go:335: production-ecs-cluster Refreshed details for sir-dbqr5wjj {
03:08:58
SpotInstanceRequestId: "sir-dbqr5wjj",
03:09:04
autoscaling.go:335: production-ecs-cluster Refreshed details for sir-dbqr5wjj {
03:09:04
SpotInstanceRequestId: "sir-dbqr5wjj",

```

panic: runtime error: invalid memory address or nil pointer dereference

19:18:10
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x7ff8a9fe3c48]

19:18:10
goroutine 53 [running]:

19:18:10
github.com/cristim/autospotting/core.(*autoScalingGroup).havingReadyToAttachSpotInstance(0xc42026a2d0, 0xc420689e01, 0x2)

19:18:10
/go/src/github.com/cristim/autospotting/core/autoscaling.go:471 +0xa08

19:18:10
github.com/cristim/autospotting/core.(*autoScalingGroup).process(0xc42026a2d0)

19:18:10
/go/src/github.com/cristim/autospotting/core/autoscaling.go:188 +0x360

19:18:10
github.com/cristim/autospotting/core.(*region).processEnabledAutoScalingGroups.func1(0xc42027dc00, 0xc420333a40, 0xc420329d40, 0x35, 0xc42027dc00, 0x0, 0x0, 0x0, 0x0, 0x0, ...)

19:18:10
/go/src/github.com/cristim/autospotting/core/region.go:321 +0x6f

19:18:10
created by github.com/cristim/autospotting/core.(*region).processEnabledAutoScalingGroups

19:18:10
/go/src/github.com/cristim/autospotting/core/region.go:323 +0x10c

```

--- a/autospotting.go
+++ b/autospotting.go
@@ -14,7 +14,7 @@ import (
 )
 
 type cfgData struct {
-	autospotting.Config
+	*autospotting.Config
 }
 
 var conf *cfgData
@@ -49,10 +49,11 @@ func init() {
 	}
 
 	conf = &cfgData{
-		autospotting.Config{
-			LogFile:    os.Stdout,
-			LogFlag:    log.Ldate | log.Ltime | log.Lshortfile,
-			MainRegion: region,
+		&autospotting.Config{
+			LogFile:         os.Stdout,
+			LogFlag:         log.Ldate | log.Ltime | log.Lshortfile,
+			MainRegion:      region,
+			SleepMultiplier: 1,
 		},
 	}
 

--- a/core/autoscaling.go
+++ b/core/autoscaling.go
@@ -476,7 +476,13 @@ func (a *autoScalingGroup) havingReadyToAttachSpotInstance() (*string, bool) {
 	debug.Println(instData)
 
 	if instData == nil || instData.LaunchTime == nil {
-		logger.Println("Apparently", *spotInstanceID, "is no longer running, moving on...")
+		logger.Println("Apparently", *spotInstanceID, "is no longer running, ",
+			"cancelling the spot instance request which created it...")
+
+		a.region.services.ec2.CancelSpotInstanceRequests(
+			&ec2.CancelSpotInstanceRequestsInput{
+				SpotInstanceRequestIds: []*string{activeSpotInstanceRequest.SpotInstanceRequestId},
+			})
 		return nil, true
 	}
 

--- a/core/autoscaling.go
+++ b/core/autoscaling.go
@@ -501,7 +501,8 @@ func (a *autoScalingGroup) havingReadyToAttachSpotInstance() (*string, bool) {
 	return spotInstanceID, false
 }
 
-func (a *autoScalingGroup) launchCheapestSpotInstance(azToLaunchIn *string) error {
+func (a *autoScalingGroup) launchCheapestSpotInstance(
+	azToLaunchIn *string) error {
 
 	if azToLaunchIn == nil {
 		logger.Println("Can't launch instances in any AZ, nothing to do here...")
@@ -560,7 +561,8 @@ func (a *autoScalingGroup) loadSpotInstanceRequest(
 
 func (a *autoScalingGroup) bidForSpotInstance(
 	ls *ec2.RequestSpotLaunchSpecification,
-	price float64) error {
+	price float64,
+) error {
 
 	svc := a.region.services.ec2
 

--- a/core/autoscaling_test.go
+++ b/core/autoscaling_test.go
@@ -416,7 +416,7 @@ func TestLoadDefaultConf(t *testing.T) {
 	}{
 		{name: "No configuration given",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     0,
 					MinOnDemandPercentage: 0.0,
 				},
@@ -428,7 +428,7 @@ func TestLoadDefaultConf(t *testing.T) {
 		},
 		{name: "Percentage value out of range (0-100)",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     0,
 					MinOnDemandPercentage: 142.2,
 				},
@@ -440,7 +440,7 @@ func TestLoadDefaultConf(t *testing.T) {
 		},
 		{name: "Percentage value out of range - negative (0-100)",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     0,
 					MinOnDemandPercentage: -22.2,
 				},
@@ -452,7 +452,7 @@ func TestLoadDefaultConf(t *testing.T) {
 		},
 		{name: "Percentage equals 33.0%",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     0,
 					MinOnDemandPercentage: 33.0,
 				},
@@ -470,7 +470,7 @@ func TestLoadDefaultConf(t *testing.T) {
 		},
 		{name: "Percentage equals 75.0%",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     0,
 					MinOnDemandPercentage: 75.0,
 				},
@@ -488,7 +488,7 @@ func TestLoadDefaultConf(t *testing.T) {
 		},
 		{name: "Percentage equals 100.0%",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     0,
 					MinOnDemandPercentage: 100,
 				},
@@ -506,7 +506,7 @@ func TestLoadDefaultConf(t *testing.T) {
 		},
 		{name: "Number passed out of range (negative)",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     -4,
 					MinOnDemandPercentage: 0,
 				},
@@ -518,7 +518,7 @@ func TestLoadDefaultConf(t *testing.T) {
 		},
 		{name: "Number superior to ASG size",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     50,
 					MinOnDemandPercentage: 0,
 				},
@@ -536,7 +536,7 @@ func TestLoadDefaultConf(t *testing.T) {
 		},
 		{name: "Number is valid 1",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     1,
 					MinOnDemandPercentage: 0,
 				},
@@ -554,7 +554,7 @@ func TestLoadDefaultConf(t *testing.T) {
 		},
 		{name: "Number has priority on percentage value",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     2,
 					MinOnDemandPercentage: 75,
 				},
@@ -573,7 +573,7 @@ func TestLoadDefaultConf(t *testing.T) {
 		},
 		{name: "Number is invalid so percentage value is used",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     -20,
 					MinOnDemandPercentage: 75.0,
 				},
@@ -592,7 +592,7 @@ func TestLoadDefaultConf(t *testing.T) {
 		},
 		{name: "Both number and percentage are invalid",
 			region: &region{
-				conf: Config{
+				conf: &Config{
 					MinOnDemandNumber:     -10,
 					MinOnDemandPercentage: 142.2,
 				},
@@ -1437,6 +1437,7 @@ func TestBidForSpotInstance(t *testing.T) {
 			rsls: &ec2.RequestSpotLaunchSpecification{},
 			regionASG: &region{
 				instances: makeInstances(),
+				conf:      &Config{},
 				services: connections{
 					ec2: mockEC2{
 						rsierr: nil,
@@ -1465,6 +1466,7 @@ func TestBidForSpotInstance(t *testing.T) {
 			rsls: &ec2.RequestSpotLaunchSpecification{},
 			regionASG: &region{
 				instances: makeInstances(),
+				conf:      &Config{},
 				services: connections{
 					ec2: mockEC2{
 						rsierr: errors.New("requestSpot"),
@@ -1493,6 +1495,7 @@ func TestBidForSpotInstance(t *testing.T) {
 			rsls: &ec2.RequestSpotLaunchSpecification{},
 			regionASG: &region{
 				instances: makeInstances(),
+				conf:      &Config{},
 				services: connections{
 					ec2: mockEC2{
 						rsierr: nil,
@@ -1521,6 +1524,7 @@ func TestBidForSpotInstance(t *testing.T) {
 			rsls: &ec2.RequestSpotLaunchSpecification{},
 			regionASG: &region{
 				instances: makeInstances(),
+				conf:      &Config{},
 				services: connections{
 					ec2: mockEC2{
 						rsierr: nil,
@@ -1548,6 +1552,7 @@ func TestBidForSpotInstance(t *testing.T) {
 		{name: "err during describe spot instance request",
 			rsls: &ec2.RequestSpotLaunchSpecification{},
 			regionASG: &region{
+				conf:      &Config{},
 				instances: makeInstances(),
 				services: connections{
 					ec2: mockEC2{
@@ -2376,11 +2381,11 @@ func TestReplaceOnDemandInstanceWithSpot(t *testing.T) {
 	tests := []struct {
 		name     string
 		asg      *autoScalingGroup
-		spotId   *string
+		spotID   *string
 		expected error
 	}{
 		{name: "OnDemand is replaced by spot instance - min/max/des identical",
-			spotId:   aws.String("spot-running"),
+			spotID:   aws.String("spot-running"),
 			expected: nil,
 			asg: &autoScalingGroup{
 				name: "test-asg",
@@ -2499,7 +2504,7 @@ func TestReplaceOnDemandInstanceWithSpot(t *testing.T) {
 			},
 		},
 		{name: "OnDemand is replaced by spot instance - min/max/des different",
-			spotId:   aws.String("spot-running"),
+			spotID:   aws.String("spot-running"),
 			expected: nil,
 			asg: &autoScalingGroup{
 				name: "test-asg",
@@ -2566,7 +2571,7 @@ func TestReplaceOnDemandInstanceWithSpot(t *testing.T) {
 			},
 		},
 		{name: "no spot instances found in region",
-			spotId:   aws.String("spot-not-found"),
+			spotID:   aws.String("spot-not-found"),
 			expected: errors.New("couldn't find spot instance to use"),
 			asg: &autoScalingGroup{
 				name: "test-asg",
@@ -2617,7 +2622,7 @@ func TestReplaceOnDemandInstanceWithSpot(t *testing.T) {
 			},
 		},
 		{name: "no OnDemand instances found in asg",
-			spotId:   aws.String("spot-running"),
+			spotID:   aws.String("spot-running"),
 			expected: errors.New("couldn't find ondemand instance to replace"),
 			asg: &autoScalingGroup{
 				name: "test-asg",
@@ -2664,7 +2669,7 @@ func TestReplaceOnDemandInstanceWithSpot(t *testing.T) {
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
-			returned := tt.asg.replaceOnDemandInstanceWithSpot(tt.spotId)
+			returned := tt.asg.replaceOnDemandInstanceWithSpot(tt.spotID)
 			CheckErrors(t, returned, tt.expected)
 		})
 	}

--- a/core/config.go
+++ b/core/config.go
@@ -2,6 +2,7 @@ package autospotting
 
 import (
 	"io"
+	"time"
 
 	"github.com/cristim/ec2-instances-info"
 )
@@ -23,4 +24,9 @@ type Config struct {
 	MinOnDemandNumber     int64
 	MinOnDemandPercentage float64
 	Regions               string
+
+	// This is only here for tests, where we want to be able to somehow mock
+	// time.Sleep without actually sleeping. While testing it defaults to 0 (which won't sleep at all), in
+	// real-world usage it's expected to be set to 1
+	SleepMultiplier time.Duration
 }

--- a/core/instance.go
+++ b/core/instance.go
@@ -264,7 +264,7 @@ func (i *instance) getCheapestCompatibleSpotInstanceType() (string, error) {
 	return chosenSpotType, fmt.Errorf("No cheaper spot instance types could be found")
 }
 
-func (i *instance) tag(tags []*ec2.Tag, maxIter int, sleepFunc func(d time.Duration)) error {
+func (i *instance) tag(tags []*ec2.Tag, maxIter int) error {
 	var (
 		n   int
 		err error
@@ -295,7 +295,7 @@ func (i *instance) tag(tags []*ec2.Tag, maxIter int, sleepFunc func(d time.Durat
 			"Failed to create tags for the spot instance", *i.InstanceId, err.Error())
 		logger.Println(i.region.name,
 			"Sleeping for 5 seconds before retrying")
-		sleepFunc(5 * time.Second)
+		time.Sleep(5 * time.Second * i.region.conf.SleepMultiplier)
 	}
 	return err
 }

--- a/core/instance_test.go
+++ b/core/instance_test.go
@@ -4,7 +4,6 @@ import (
 	"errors"
 	"reflect"
 	"testing"
-	"time"
 
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/autoscaling"
@@ -988,6 +987,7 @@ func TestTag(t *testing.T) {
 				},
 				region: &region{
 					name: "test",
+					conf: &Config{},
 					services: connections{
 						ec2: mockEC2{
 							cterr: nil,
@@ -1006,6 +1006,7 @@ func TestTag(t *testing.T) {
 				},
 				region: &region{
 					name: "test",
+					conf: &Config{},
 					services: connections{
 						ec2: mockEC2{
 							cterr: errors.New("no tags with error"),
@@ -1027,6 +1028,7 @@ func TestTag(t *testing.T) {
 				},
 				region: &region{
 					name: "test",
+					conf: &Config{},
 					services: connections{
 						ec2: mockEC2{
 							cterr: nil,
@@ -1048,6 +1050,7 @@ func TestTag(t *testing.T) {
 				},
 				region: &region{
 					name: "test",
+					conf: &Config{},
 					services: connections{
 						ec2: mockEC2{
 							cterr: errors.New("tags with error"),
@@ -1059,10 +1062,8 @@ func TestTag(t *testing.T) {
 		},
 	}
 
-	mockedSleep := func(time.Duration) {}
-
 	for _, tt := range tests {
-		err := tt.inst.tag(tt.tags, 1, mockedSleep)
+		err := tt.inst.tag(tt.tags, 1)
 		CheckErrors(t, err, tt.expectedError)
 	}
 }

--- a/core/main.go
+++ b/core/main.go
@@ -17,11 +17,11 @@ var logger, debug *log.Logger
 // Run starts processing all AWS regions looking for AutoScaling groups
 // enabled and taking action by replacing more pricy on-demand instances with
 // compatible and cheaper spot instances.
-func Run(cfg Config) {
+func Run(cfg *Config) {
 
 	setupLogging(cfg)
 
-	debug.Println(cfg)
+	debug.Println(*cfg)
 
 	// use this only to list all the other regions
 	ec2Conn := connectEC2(cfg.MainRegion)
@@ -38,10 +38,10 @@ func Run(cfg Config) {
 }
 
 func disableLogging() {
-	setupLogging(Config{LogFile: ioutil.Discard})
+	setupLogging(&Config{LogFile: ioutil.Discard})
 }
 
-func setupLogging(cfg Config) {
+func setupLogging(cfg *Config) {
 	logger = log.New(cfg.LogFile, "", cfg.LogFlag)
 
 	if os.Getenv("AUTOSPOTTING_DEBUG") == "true" {
@@ -54,7 +54,7 @@ func setupLogging(cfg Config) {
 
 // processAllRegions iterates all regions in parallel, and replaces instances
 // for each of the ASGs tagged with 'spot-enabled=true'.
-func processRegions(regions []string, cfg Config) {
+func processRegions(regions []string, cfg *Config) {
 
 	var wg sync.WaitGroup
 

--- a/core/mock_test.go
+++ b/core/mock_test.go
@@ -55,6 +55,10 @@ type mockEC2 struct {
 	// Describe Regions
 	dro   *ec2.DescribeRegionsOutput
 	drerr error
+
+	//Cancel Spot instance request
+	csiro   *ec2.CancelSpotInstanceRequestsOutput
+	csirerr error
 }
 
 func (m mockEC2) CreateTags(in *ec2.CreateTagsInput) (*ec2.CreateTagsOutput, error) {
@@ -83,6 +87,10 @@ func (m mockEC2) TerminateInstances(*ec2.TerminateInstancesInput) (*ec2.Terminat
 
 func (m mockEC2) RequestSpotInstances(*ec2.RequestSpotInstancesInput) (*ec2.RequestSpotInstancesOutput, error) {
 	return m.rsio, m.rsierr
+}
+
+func (m mockEC2) CancelSpotInstanceRequests(*ec2.CancelSpotInstanceRequestsInput) (*ec2.CancelSpotInstanceRequestsOutput, error) {
+	return m.csiro, m.csirerr
 }
 
 func (m mockEC2) DescribeRegions(*ec2.DescribeRegionsInput) (*ec2.DescribeRegionsOutput, error) {

--- a/core/region.go
+++ b/core/region.go
@@ -17,7 +17,7 @@ import (
 type region struct {
 	name string
 
-	conf Config
+	conf *Config
 	// The key in this map is the instance type.
 	instanceTypeInformation map[string]instanceTypeInformation
 
@@ -146,7 +146,7 @@ func (r *region) addInstance(inst *ec2.Instance) {
 	})
 }
 
-func (r *region) determineInstanceTypeInformation(cfg Config) {
+func (r *region) determineInstanceTypeInformation(cfg *Config) {
 
 	r.instanceTypeInformation = make(map[string]instanceTypeInformation)
 

--- a/core/region_test.go
+++ b/core/region_test.go
@@ -91,7 +91,7 @@ func Test_region_enabled(t *testing.T) {
 		t.Run(tt.name, func(t *testing.T) {
 			r := &region{
 				name: tt.region,
-				conf: Config{
+				conf: &Config{
 					Regions: tt.allowed,
 				},
 			}

--- a/core/spot_instance_request.go
+++ b/core/spot_instance_request.go
@@ -1,6 +1,7 @@
 package autospotting
 
 import (
+	"errors"
 	"time"
 
 	"github.com/aws/aws-sdk-go/aws"
@@ -70,24 +71,33 @@ func (s *spotInstanceRequest) waitForAndTagSpotInstance() error {
 
 func (s *spotInstanceRequest) tag(asgName string) error {
 	svc := s.region.services.ec2
-
-	_, err := svc.CreateTags(&ec2.CreateTagsInput{
-		Resources: []*string{s.SpotInstanceRequestId},
-		Tags: []*ec2.Tag{
-			{
-				Key:   aws.String("launched-for-asg"),
-				Value: aws.String(asgName),
-			},
+	tags := []*ec2.Tag{
+		{
+			Key:   aws.String("launched-for-asg"),
+			Value: aws.String(asgName),
 		},
-	})
+	}
 
-	if err != nil {
-		// Print the error, cast err to awserr.Error to get the Code and
-		// Message from an error.
-		logger.Println(asgName,
-			"Failed to create tags for the spot instance request",
-			err.Error())
-		return err
+	for count, err := 0, errors.New("dummy"); err != nil; _, err = svc.CreateTags(&ec2.CreateTagsInput{
+		Resources: []*string{s.SpotInstanceRequestId},
+		Tags:      tags,
+	}) {
+
+		if err != nil {
+			if count > 10 {
+				logger.Println(asgName,
+					"Failed to create tags for the spot instance request",
+					err.Error())
+				return err
+
+			}
+			logger.Println(asgName,
+				"Failed to create tags for the spot instance request",
+				*s.SpotInstanceRequestId, "retrying in 5 seconds...")
+			count = count + 1
+			time.Sleep(5 * time.Second)
+
+		}
 	}
 
 	logger.Println(asgName, "successfully tagged spot instance request",

--- a/core/spot_instance_request.go
+++ b/core/spot_instance_request.go
@@ -62,7 +62,7 @@ func (s *spotInstanceRequest) waitForAndTagSpotInstance() error {
 	i := s.region.instances.get(*spotInstanceID)
 
 	if i != nil {
-		i.tag(tags, defaultTimeout, time.Sleep)
+		i.tag(tags, defaultTimeout)
 	} else {
 		logger.Println(s.asg.name, "new spot instance", *spotInstanceID, "has disappeared")
 	}
@@ -83,11 +83,24 @@ func (s *spotInstanceRequest) tag(asgName string) error {
 		Tags:      tags,
 	}) {
 
+		// after failing to tag it for 10 retries, terminate its instance and cancel
+		// the spot instance request in order to avoid any orphaned instances
+		// created by spot requests which failed to be tagged.
 		if err != nil {
 			if count > 10 {
 				logger.Println(asgName,
-					"Failed to create tags for the spot instance request",
-					err.Error())
+					"Failed to create tags for the spot instance request after 10 retries",
+					"cancelling the spot instance request, error: ", err.Error())
+				s.reload()
+
+				svc.TerminateInstances(&ec2.TerminateInstancesInput{
+					InstanceIds: []*string{s.InstanceId},
+				})
+
+				svc.CancelSpotInstanceRequests(
+					&ec2.CancelSpotInstanceRequestsInput{
+						SpotInstanceRequestIds: []*string{s.SpotInstanceRequestId},
+					})
 				return err
 
 			}
@@ -95,7 +108,7 @@ func (s *spotInstanceRequest) tag(asgName string) error {
 				"Failed to create tags for the spot instance request",
 				*s.SpotInstanceRequestId, "retrying in 5 seconds...")
 			count = count + 1
-			time.Sleep(5 * time.Second)
+			time.Sleep(5 * time.Second * s.region.conf.SleepMultiplier)
 
 		}
 	}
@@ -104,4 +117,23 @@ func (s *spotInstanceRequest) tag(asgName string) error {
 		*s.SpotInstanceRequestId)
 
 	return nil
+}
+
+func (s *spotInstanceRequest) reload() error {
+
+	resp, err := s.region.services.ec2.DescribeSpotInstanceRequests(
+		&ec2.DescribeSpotInstanceRequestsInput{
+			SpotInstanceRequestIds: []*string{s.SpotInstanceRequestId},
+		})
+
+	if err != nil {
+		return err
+	}
+
+	if resp != nil && len(resp.SpotInstanceRequests) == 1 {
+		s.SpotInstanceRequest = resp.SpotInstanceRequests[0]
+		return nil
+	}
+	return errors.New("spot instance request not found")
+
 }

--- a/core/spot_instance_request_test.go
+++ b/core/spot_instance_request_test.go
@@ -22,6 +22,7 @@ func Test_waitForAndTagSpotInstance(t *testing.T) {
 					SpotInstanceRequestId: aws.String(""),
 				},
 				region: &region{
+					conf: &Config{},
 					services: connections{
 						ec2: mockEC2{
 							wusirferr: errors.New(""),
@@ -40,6 +41,7 @@ func Test_waitForAndTagSpotInstance(t *testing.T) {
 					SpotInstanceRequestId: aws.String(""),
 				},
 				region: &region{
+					conf: &Config{},
 					services: connections{
 						ec2: mockEC2{
 							dsiro: &ec2.DescribeSpotInstanceRequestsOutput{
@@ -66,6 +68,7 @@ func Test_waitForAndTagSpotInstance(t *testing.T) {
 					SpotInstanceRequestId: aws.String(""),
 				},
 				region: &region{
+					conf: &Config{},
 					services: connections{
 						ec2: mockEC2{
 							dsiro: &ec2.DescribeSpotInstanceRequestsOutput{
@@ -108,6 +111,7 @@ func Test_tag(t *testing.T) {
 					SpotInstanceRequestId: aws.String(""),
 				},
 				region: &region{
+					conf: &Config{},
 					services: connections{
 						ec2: mockEC2{
 							cterr: errors.New(""),
@@ -125,6 +129,7 @@ func Test_tag(t *testing.T) {
 					SpotInstanceRequestId: aws.String(""),
 				},
 				region: &region{
+					conf: &Config{},
 					services: connections{
 						ec2: mockEC2{
 							cterr: nil,

--- a/core/spot_price_test.go
+++ b/core/spot_price_test.go
@@ -146,12 +146,12 @@ func Test_fetch(t *testing.T) {
 						dspho: &ec2.DescribeSpotPriceHistoryOutput{
 							SpotPriceHistory: []*ec2.SpotPrice{},
 						},
-						dspherr: errors.New("Error!"),
+						dspherr: errors.New("error"),
 					},
 				},
 			},
 			data: []*ec2.SpotPrice{},
-			err:  errors.New("Error!"),
+			err:  errors.New("error"),
 		},
 		{
 			name: "ok",

